mrmustard/lab_dev/circuit_components_utils/branch_and_bound.py

```diff
@@ -0,0 +1,444 @@
+
+        Returns:
+            int: contraction cost in approx FLOPS
+
+        """
+        idxA, idxB = self.wires.contracted_indices(other.wires)
+        m = len(idxA)  # same as len(idxB)
+        nA, nB = len(self.shape) - m, len(other.shape) - m
+
+        if self.repr == "Bargmann" and other.repr == "Bargmann":
+            cost = (  # +1s to include vector part)
+                m * m * m  # M inverse
+                + (m + 1) * m * nA  # left matmul
+                + (m + 1) * m * nB  # right matmul
+                + (m + 1) * m  # addition
+                + m * m * m  # determinant of M
+            )
+        else:  # otherwise need to use fock cost
+            prod_A = np.prod([s for i, s in enumerate(self.shape) if i not in idxA])
+            prod_B = np.prod([s for i, s in enumerate(other.shape) if i not in idxB])
+            prod_contracted = np.prod(
+                [min(self.shape[i], other.shape[j]) for i, j in zip(idxA, idxB)]
+            )
+            cost = (
+                prod_A * prod_B * prod_contracted  # matmul
+                + np.prod(self.shape) * (self.repr == "Bargmann")  # conversion
+                + np.prod(other.shape) * (other.repr == "Bargmann")  # conversion
+            )
+        return int(cost)
+
+    def __matmul__(self, other) -> GraphComponent:
+        r"""
+        Returns the contracted GraphComponent.
+        """
+        new_wires, perm = self.wires @ other.wires
+        idxA, idxB = self.wires.contracted_indices(other.wires)
+        shape_A = [n for i, n in enumerate(self.shape) if i not in idxA]
+        shape_B = [n for i, n in enumerate(other.shape) if i not in idxB]
+        shape = shape_A + shape_B
+        new_shape = [shape[p] for p in perm]
+        new_component = GraphComponent(
+            "Bargmann" if self.repr == other.repr == "Bargmann" else "Fock",
+            new_wires,
+            new_shape,
+            f"({self.name}@{other.name})",
+            self.contraction_cost(other) + self.cost + other.cost,
+        )
+        return new_component
+
+    def __repr__(self):
+        return f"{self.name}({self.shape}, {self.wires})"
+
+
+class Graph(nx.DiGraph):
+    r"""
+    Power pack for nx.DiGraph with additional attributes and methods.
+    """
+
+    def __init__(self, solution: tuple[Edge, ...] = (), costs: tuple[int, ...] = ()):
+        super().__init__()
+        self.solution = solution
+        self.costs = costs
+
+    @property
+    def cost(self) -> int:
+        r"""
+        Returns the total cost of the graph.
+        """
+        return sum(self.costs)
+
+    def __lt__(self, other: Graph) -> bool:
+        return self.cost < other.cost
+
+    def __hash__(self) -> int:
+        return hash(
+            tuple(self.nodes)
+            + tuple(self.edges)
+            + tuple(self.solution)
+            + tuple(sum((c.shape for c in self.components()), start=[]))
+        )
+
+    def component(self, n) -> GraphComponent:
```

Comment:
Can you reorder to init -> properties -> methods -> dunder methods

Reply:
And add type hints / docs

Reply:
done